File: web/src/types/webhook.ts
export interface Webhook {
  id: string
  workspace_id: string
  url: string
  events: string[]
  is_active: boolean
  failure_count: number
  last_triggered_at?: string | null
  last_success_at?: string | null
  created_at: string
  updated_at: string
}

export interface WebhookDelivery {
  id: string
  webhook_id: string
  event: string
  payload: Record<string, unknown>
  response_status?: number | null
  response_body?: string | null
  attempts: number
  max_attempts: number
  last_attempt_at?: string | null
  completed_at?: string | null
  created_at: string
}

export interface CreateWebhookRequest {
  url: string
  events: string[]
}

export interface CreateWebhookResponse {
  webhook: Webhook
  secret: string
}

export const WEBHOOK_EVENTS = [
  { value: "link.created", label: "Link Created", category: "Links" },
  { value: "link.updated", label: "Link Updated", category: "Links" },
  { value: "link.deleted", label: "Link Deleted", category: "Links" },
  { value: "link.clicked", label: "Link Clicked", category: "Links" },
  { value: "link.expired", label: "Link Expired", category: "Links" },
  { value: "qr.created", label: "QR Code Created", category: "QR Codes" },
  { value: "qr.scanned", label: "QR Code Scanned", category: "QR Codes" },
  { value: "biopage.created", label: "Bio Page Created", category: "Bio Pages" },
  { value: "biopage.updated", label: "Bio Page Updated", category: "Bio Pages" },
  { value: "domain.added", label: "Domain Added", category: "Domains" },
  { value: "domain.verified", label: "Domain Verified", category: "Domains" },
  { value: "domain.removed", label: "Domain Removed", category: "Domains" },
  { value: "team.member_invited", label: "Member Invited", category: "Team" },
  { value: "team.member_joined", label: "Member Joined", category: "Team" },
  { value: "team.member_removed", label: "Member Removed", category: "Team" },
] as const

File: internal/models/webhook.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"github.com/link-rift/link-rift/internal/repository/sqlc"
)

// ValidWebhookEvents defines all valid webhook event types.
var ValidWebhookEvents = []string{
	"link.created",
	"link.updated",
	"link.deleted",
	"link.clicked",
	"link.expired",
	"qr.created",
	"qr.scanned",
	"biopage.created",
	"biopage.updated",
	"domain.added",
	"domain.verified",
	"domain.removed",
	"team.member_invited",
	"team.member_joined",
	"team.member_removed",
}

type Webhook struct {
	ID              uuid.UUID  `json:"id"`
	WorkspaceID     uuid.UUID  `json:"workspace_id"`
	URL             string     `json:"url"`
	Secret          string     `json:"-"`
	Events          []string   `json:"events"`
	IsActive        bool       `json:"is_active"`
	FailureCount    int32      `json:"failure_count"`
	LastTriggeredAt *time.Time `json:"last_triggered_at,omitempty"`
	LastSuccessAt   *time.Time `json:"last_success_at,omitempty"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`
}

type WebhookDelivery struct {
	ID             uuid.UUID       `json:"id"`
	WebhookID      uuid.UUID       `json:"webhook_id"`
	Event          string          `json:"event"`
	Payload        json.RawMessage `json:"payload"`
	ResponseStatus *int32          `json:"response_status,omitempty"`
	ResponseBody   *string         `json:"response_body,omitempty"`
	Attempts       int32           `json:"attempts"`
	MaxAttempts    int32           `json:"max_attempts"`
	LastAttemptAt  *time.Time      `json:"last_attempt_at,omitempty"`
	CompletedAt    *time.Time      `json:"completed_at,omitempty"`
	CreatedAt      time.Time       `json:"created_at"`
}

type WebhookEvent struct {
	Event       string          `json:"event"`
	WorkspaceID uuid.UUID       `json:"workspace_id"`
	Timestamp   time.Time       `json:"timestamp"`
	Data        json.RawMessage `json:"data"`
}

type CreateWebhookInput struct {
	URL    string   `json:"url" binding:"required,url"`
	Events []string `json:"events" binding:"required,min=1"`
}

type CreateWebhookResponse struct {
	Webhook *Webhook `json:"webhook"`
	Secret  string   `json:"secret"`
}

func WebhookFromSqlc(w sqlc.Webhook) *Webhook {
	wh := &Webhook{
		ID:           w.ID,
		WorkspaceID:  w.WorkspaceID,
		URL:          w.Url,
		Secret:       w.Secret,
		Events:       w.Events,
		IsActive:     w.IsActive,
		FailureCount: w.FailureCount,
	}
	if w.LastTriggeredAt.Valid {
		t := w.LastTriggeredAt.Time
		wh.LastTriggeredAt = &t
	}
	if w.LastSuccessAt.Valid {
		t := w.LastSuccessAt.Time
		wh.LastSuccessAt = &t
	}
	if w.CreatedAt.Valid {
		wh.CreatedAt = w.CreatedAt.Time
	}
	if w.UpdatedAt.Valid {
		wh.UpdatedAt = w.UpdatedAt.Time
	}
	return wh
}

func WebhookDeliveryFromSqlc(d sqlc.WebhookDelivery) *WebhookDelivery {
	wd := &WebhookDelivery{
		ID:          d.ID,
		WebhookID:   d.WebhookID,
		Event:       d.Event,
		Payload:     d.Payload,
		Attempts:    d.Attempts,
		MaxAttempts: d.MaxAttempts,
	}
	if d.ResponseStatus.Valid {
		v := d.ResponseStatus.Int32
		wd.ResponseStatus = &v
	}
	if d.ResponseBody.Valid {
		wd.ResponseBody = &d.ResponseBody.String
	}
	if d.LastAttemptAt.Valid {
		t := d.LastAttemptAt.Time
		wd.LastAttemptAt = &t
	}
	if d.CompletedAt.Valid {
		t := d.CompletedAt.Time
		wd.CompletedAt = &t
	}
	if d.CreatedAt.Valid {
		wd.CreatedAt = d.CreatedAt.Time
	}
	return wd
}

func IsValidWebhookEvent(event string) bool {
	for _, e := range ValidWebhookEvents {
		if e == event {
			return true
		}
	}
	return false
}
